[Merged by Bors] - feat(algebra/ring/basic): mul_{left,right}_cancel_of_non_zero_divisor

pechersky opened the pull request:
We also golf the proof that a domain is a cancel_monoid_with_zero.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

pechersky committed:
feat(algebra/ring/basic): mul_{left,right}_cancel_of_non_zero_divisor

pechersky committed:
smul_cancel

pechersky committed:
remove empty line

eric-wieser commented on src/algebra/group_ring_action.lean:
```diff
@@ -130,6 +130,15 @@ end semiring
 
 section ring
 
+lemma smul_cancel_of_non_zero_divisor {M R : Type*} [monoid M] [ring R] [distrib_mul_action M R]
```
Don't we have a typeclass for this?

eric-wieser replied:
cc @Vierkantor. I think we should at least mention `no_zero_smul_divisors` in the docstring even if we don't use it here

eric-wieser replied:
Also this is the wrong file, as this is only supposed to be about `mul_semiring_action`

pechersky replied:
The top of the file says
```
There is no separate typeclass for group acting on rings, group acting on fields, etc.
They are all grouped under `mul_semiring_action`.
```
Do you mean this should be earlier in the import hierarchy, in a file about `distrib_mul_action`?

eric-wieser commented on src/algebra/ring/basic.lean:
```diff
@@ -885,6 +885,22 @@ lemma succ_ne_self [ring α] [nontrivial α] (a : α) : a + 1 ≠ a :=
 λ h, one_ne_zero (neg_injective ((add_right_inj a).mp (by simpa [sub_eq_add_neg] using h)))
 
+lemma mul_left_cancel_of_non_zero_divisor [ring α] (k : α)
+  (h : ∀ (x : α), k * x = 0 → x = 0) {x y : α} (h' : k * x = k * y) : x = y :=
```
@adomani, this is part of your `regular` API, right?

pechersky committed:
move lemmas and docstring suggestions

Vierkantor reviewed (commented):
I'm a bit worried that we're getting too many unconnected ways to express the same thing: `is_regular x`, `x ∈ non_zero_divisors α` and, implicitly, here as well.

It looks like `algebra/regular.lean` doesn't import anything heavy, so probably the best way to deal with this is to move the `mul_{left,right}_cancel_of_non_zero_divisor` (expressed in terms of `is_{left,right}_regular`) to `regular.lean`, and then import it in `ring/basic.lean`.

pechersky committed:
refactor(algebra/regular): split out `is_regular`

One would like to import `is_regular` for rings. However, group powers
are too late in the algebra hierarchy,
so the proofs of powers of regular elements are factored
out to a separate file.

pechersky committed:
utilize regular

pechersky committed:
fix naming

pechersky committed:
Merge branch 'master' into pechersky/ring-cancel-non-zero-divisor

Vierkantor commented on src/algebra/ring/basic.lean:
```diff
@@ -908,9 +939,9 @@ instance domain.to_no_zero_divisors : no_zero_divisors α :=
 instance domain.to_cancel_monoid_with_zero : cancel_monoid_with_zero α :=
 { mul_left_cancel_of_ne_zero := λ a b c ha,
-    by { rw [← sub_eq_zero, ← mul_sub], simp [ha, sub_eq_zero] },
+     @is_regular.left _ _ _ (is_regular_of_ne_zero' ha) _ _,
```
```suggestion
    @is_regular.left _ _ _ (is_regular_of_ne_zero' ha) _ _,
```

Vierkantor commented on src/algebra/ring/basic.lean:
```diff
@@ -908,9 +939,9 @@ instance domain.to_no_zero_divisors : no_zero_divisors α :=
+     @is_regular.left _ _ _ (is_regular_of_ne_zero' ha) _ _,
   mul_right_cancel_of_ne_zero := λ a b c hb,
-    by { rw [← sub_eq_zero, ← sub_mul], simp [hb, sub_eq_zero] },
+     @is_regular.right _ _ _ (is_regular_of_ne_zero' hb) _ _,
```
```suggestion
    @is_regular.right _ _ _ (is_regular_of_ne_zero' hb) _ _,
```

Vierkantor reviewed (approved):
Thanks!

bors d+

pechersky committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>
